cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -0,0 +1,507 @@
+
+namespace cudf::io::parquet::detail {
+
+namespace {
+
+constexpr int decode_block_size = 128;
+constexpr int rolling_buf_size  = decode_block_size * 2;
+// the required number of runs in shared memory we will need to provide the
+// rle_stream object
+constexpr int rle_run_buffer_size = rle_stream_required_run_buffer_size<decode_block_size>();
+
+template <bool nullable, typename level_t, typename state_buf>
+static __device__ int gpuUpdateValidityOffsetsAndRowIndicesFlat(int32_t target_value_count,
+                                                                page_state_s* s,
+                                                                state_buf* sb,
+                                                                level_t const* const def,
+                                                                int t,
+                                                                bool has_no_nulls)
+{
+  constexpr int num_warps      = decode_block_size / cudf::detail::warp_size;
+  constexpr int max_batch_size = num_warps * cudf::detail::warp_size;
+
+  auto& ni = s->nesting_info[0];
+
+  // how many (input) values we've processed in the page so far
+  int value_count = s->input_value_count;
+  int valid_count = ni.valid_count;
+
+  // cap by last row so that we don't process any rows past what we want to output.
+  int const first_row = s->first_row;
+  int const last_row  = first_row + s->num_rows;
+  target_value_count  = min(target_value_count, last_row);
+
+  int const valid_map_offset      = ni.valid_map_offset;
+  int const row_index_lower_bound = s->row_index_lower_bound;
+
+  __syncthreads();
+
+  while (value_count < target_value_count) {
+    int const batch_size = min(max_batch_size, target_value_count - value_count);
+
+    // definition level. only need to process for nullable columns
+    int d = 0;
+    if constexpr (nullable) {
+      if (has_no_nulls) {
+        d = 1;
+      } else {
+        d = t < batch_size
+              ? static_cast<int>(def[rolling_index<state_buf::nz_buf_size>(value_count + t)])
+              : -1;
+      }
+    }
+
+    int const thread_value_count = t + 1;
+    int const block_value_count  = batch_size;
+
+    // compute our row index, whether we're in row bounds, and validity
+    int const row_index     = (thread_value_count + value_count) - 1;
+    int const in_row_bounds = (row_index >= row_index_lower_bound) && (row_index < last_row);
+    int is_valid;
+    if constexpr (nullable) {
+      is_valid = ((d > 0) && in_row_bounds) ? 1 : 0;
+    } else {
+      is_valid = in_row_bounds;
+    }
+
+    // thread and block validity count
+    int thread_valid_count, block_valid_count;
+    if constexpr (nullable) {
+      using block_scan = cub::BlockScan<int, decode_block_size>;
+      __shared__ typename block_scan::TempStorage scan_storage;
+      block_scan(scan_storage).InclusiveSum(is_valid, thread_valid_count, block_valid_count);
+      __syncthreads();
```

Comment:
```suggestion
      if (has_no_nulls) {
        value_count += block_value_count;
        valid_count += block_valid_count;
      } else {
        using block_scan = cub::BlockScan<int, decode_block_size>;
        __shared__ typename block_scan::TempStorage scan_storage;
        block_scan(scan_storage).InclusiveSum(is_valid, thread_valid_count, block_valid_count);
        __syncthreads();
      }
```
This is what I was trying to say 😅

Reply:
Thanks @etseidl. I'd like for us to postpone this until after this PR is done. I need to be able to compute `block_valid_count` and it would be a bit of work to figure out and get right, and I am nervous of making more changes right now. That said, I am excited to have changes like this make it in the near future. @nvdbaranec fyi. I'll file a follow on issue.

Reply:
Agreed on the followup PR.

Reply:
Follow on filed. https://github.com/rapidsai/cudf/issues/15266